### PR TITLE
add Gemma2 support to MaxText 

### PR DESCRIPTION
**Pull request by ZhaoyueCheng**
add Gemma2 support to MaxText on top of [PR](https://github.com/google/maxtext/pull/804)

+ add `gemma` Decoder block which merges [local_sliding_attention, global_attention] combination into one Decoder layer with `post_attn_norm`, `post_ffw_norm` support
+ add `convert_gemma2_chkpt.py` to convert checkpoint from Gemma2 architecture
   + merges one [local_sliding_attention, global_attention] combination into one Decoder layer
   + adds `post_attn_norm`, `post_ffw_norm`, `transpose_gating_einsum` and `query_pre_attn_scalar` support which is new from Gemma1
+ enabling `post_attn_norm`, `post_ffw_norm` in the Gemma2 Decoder architecture
+ add `final_logits_soft_cap` support
+ add `gemma2-2b` `gemma2-9b` `gemma2-27b` config yaml file and 2b end_to_end test script and test golden logits dumped from flax

### CODE REVIEW

File: MaxText/configs/base.yml [hunk 1]
```diff
@@ -137,11 +137,13 @@ param_scan_axis: 1
 
 # The attention parameter dictates the specific algorithm/methodology used to compute the attention scores
 attention: 'autoselected' # Supported attention: autoselected, dot_product, flash, cudnn_flash_te
-
-# The attention_type parameter determines the variants of attention, e.g. global or local_sliding
```
Comment (gagika) [inline MaxText/configs/base.yml:141]: Could you keep those comments?

File: MaxText/configs/models/gemma2-9b.yml [hunk 2]
```diff
@@ -12,21 +12,22 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-# model config for gemma2-9B
```
Comment (gagika) [inline MaxText/configs/models/gemma2-9b.yml:15]: Please update gemma2-9B

File: end_to_end/tpu/gemma2/9b/1_test_gemma.sh [hunk 3]
```diff
@@ -1,46 +0,0 @@
-#!/bin/bash
```
Comment (gagika) [inline end_to_end/tpu/gemma2/9b/1_test_gemma.sh]: Why was it deleted?

File: end_to_end/tpu/gemma2/9b/2_test_gemma.sh [hunk 4]
```diff
@@ -1,43 +0,0 @@
-#!/bin/bash
```
Comment (gagika) [inline end_to_end/tpu/gemma2/9b/2_test_gemma.sh]: Why was this file deleted?

File: MaxText/pyconfig.py [hunk 5]
```diff
@@ -147,6 +147,9 @@ def validate_model_name(s: str) -> bool:
       "llama3-70b",
       "mistral-7b",
       "mixtral-8x7b",
+      "gemma2-2b",
+      "gemma2-9b",
+      "gemma2-27b"
```
Comment (gagika) [inline MaxText/pyconfig.py:152]: Can you move it below gemma-* ?
Also there is gemma2-9B in the list.

File: MaxText/scratch_code/golden_gemma2-2b_export-flax.ipynb [hunk 6]
```diff
@@ -0,0 +1,331 @@
+{
```
Comment (gagika) [inline MaxText/scratch_code/golden_gemma2-2b_export-flax.ipynb:1]: Do we want to open source all the notebooks?

**Review by gagika — COMMENTED** [review]
Some quick comments

File: end_to_end/tpu/gemma2/9b/1_test_gemma.sh (hunk 3 above)
Comment (ZhaoyueCheng) [reply end_to_end/tpu/gemma2/9b/1_test_gemma.sh]: updated, I think it got deleted somehow during the rebase

File: end_to_end/tpu/gemma2/9b/2_test_gemma.sh (hunk 4 above)
Comment (ZhaoyueCheng) [reply end_to_end/tpu/gemma2/9b/2_test_gemma.sh]: updated, I think it got deleted somehow during the rebase

File: MaxText/configs/models/gemma2-9b.yml (hunk 2 above)
Comment (ZhaoyueCheng) [reply MaxText/configs/models/gemma2-9b.yml:15]: updated

File: MaxText/pyconfig.py (hunk 5 above)
Comment (ZhaoyueCheng) [reply MaxText/pyconfig.py]: updated

File: MaxText/configs/base.yml (hunk 1 above)
Comment (ZhaoyueCheng) [reply MaxText/configs/base.yml:141]: updated

File: MaxText/scratch_code/golden_gemma2-2b_export-flax.ipynb (hunk 6 above)
Comment (ZhaoyueCheng) [reply MaxText/scratch_code/golden_gemma2-2b_export-flax.ipynb:1]: I think we had the original [gemma export notebook](https://github.com/google/maxtext/blob/main/MaxText/scratch_code/golden_gemma-2b_export.ipynb) and other notebooks (llama, mixtral, etc) open sourced so added the gemma2 notebook to scratch_code folder as well with other open source notebooks

**Comment by salrowili** [comment]
Please note that if we merge local_sliding_attention, global_attention and assigned half the decoder layer to "base_num_decoder_layers", the TFLOPS calculation will report half the actual TFLOPS . This is due to how maxtext_utils calcuate the TFLOPS
https://github.com/google/maxtext/blob/644eb87ae90dd8b210ce17f1c16ca7a54e80fceb/MaxText/maxtext_utils.py#L139

**Comment by gobbleturk** [comment]
> Please note that if we merge local_sliding_attention, global_attention and assigned half the decoder layer to "base_num_decoder_layers", the TFLOPS calculation will report half the actual TFLOPS . This is due to how maxtext_utils calcuate the TFLOPS
> 
> https://github.com/google/maxtext/blob/644eb87ae90dd8b210ce17f1c16ca7a54e80fceb/MaxText/maxtext_utils.py#L139

Good catch! Thank you for noting this

**Comment by ZhaoyueCheng** [comment]
> Please note that if we merge local_sliding_attention, global_attention and assigned half the decoder layer to "base_num_decoder_layers", the TFLOPS calculation will report half the actual TFLOPS . This is due to how maxtext_utils calcuate the TFLOPS
> 
> https://github.com/google/maxtext/blob/644eb87ae90dd8b210ce17f1c16ca7a54e80fceb/MaxText/maxtext_utils.py#L139

updated, thanks for the comment!

File: MaxText/layers/gemma2.py [hunk 7]
```diff
@@ -0,0 +1,246 @@
+"""
+Copyright 2023 Google LLC
```
Comment (khatwanimohit) [inline MaxText/layers/gemma2.py:2]: nit: Update the copyright to 2024

File: MaxText/pyconfig.py [hunk 8]
```diff
@@ -149,7 +149,9 @@ def validate_model_name(s: str) -> bool:
       "mixtral-8x7b",
       "gemma-7b",
       "gemma-2b",
+      "gemma2-2b",
       "gemma2-9b",
+      "gemma2-27b",
```
Comment (khatwanimohit) [inline MaxText/pyconfig.py:154]: nit: Should we remove this key for now until we add the gemma2-27b config yml file ? Just don't want people to assume gemma-27b is supported now.

**Review by khatwanimohit — APPROVED** [review]
Thanks for adding this! Left a few nits. 

QQ: Are we going to add logit checker for Gemma9B in a separate PR? Since you have the code for golden logits here, can you also add golden logits jsonl file for Gemma2-9B in this PR

File: MaxText/pyconfig.py (hunk 8 above)
Comment (ZhaoyueCheng) [reply MaxText/pyconfig.py:154]: added gemma2-27b config yaml file, thanks for the review!

File: MaxText/layers/gemma2.py (hunk 7 above)
Comment (ZhaoyueCheng) [reply MaxText/layers/gemma2.py]: updated, thanks for the review!

File: MaxText/maxtext_utils.py [hunk 9]
```diff
@@ -139,6 +139,11 @@ def calculate_tflops_training_per_device(config, log=True):
       attention_flops * config.num_decoder_layers * 3 / 10**12
   )
 
+  # in gemma2 we combine [local attention, global attention] into one decoder block, so need to multiply the
```
Comment (gobbleturk) [inline MaxText/maxtext_utils.py:142]: Don't we also need to multiply the FF by 2 as well since there are 2 FF per gemma2 layer? I'm not sure the best way to handle this, I would rather not have decoder-block specific logic, however here it is necessary. Can we move it to its own subfunction, e.g. "if config.decoer_Block == gemma2: return calculate_gemma2_flops(config)
else:
calculate standard_flops(config)"

If you can think of a better solution let me know

File: MaxText/pyconfig.py (hunk 8 above)
Comment (gobbleturk) [reply MaxText/pyconfig.py:154]: Have we tested golden logits for the gemma2-27b as well?

**Review by gobbleturk — CHANGES_REQUESTED** [review]
Overall looks very good, thank you for adding this! Just want to clean up the flops calculation.

File: MaxText/maxtext_utils.py (hunk 9 above)
Comment (gobbleturk) [reply MaxText/maxtext_utils.py:142]: I think we should have calculate_gemma2_flops defined inside of the gemma2 decoder block definition actually, and call it from max_utils. Ideally some inheritance with a default method inherited that works for most models, but can be overriden, we max_utils will call decoder_block.calculate_tflops()...

We can do in the inheritance later, but for this PR we should at least calculate the flops accurately ideally in its own method in either max_utils or the gemma2 decoder block

File: MaxText/pyconfig.py (hunk 8 above)
Comment (ZhaoyueCheng) [reply MaxText/pyconfig.py]: added the 2b and 9b, for 9b I tested it on `JAX_PLATFORMS=cpu` with `dtype=float32` locally and it works, for 27b it OOMs even on CPU with scanned checkpoint so removing it from the `pyconfig` file for now

File: MaxText/maxtext_utils.py (hunk 9 above)
Comment (ZhaoyueCheng) [reply MaxText/maxtext_utils.py]: Thanks for the catch! while refactoring I found since most of the code in the new `calculate_gemma2_flops` are the same I use a local variable `num_decoder_layers` and multiply it by 2 when using the `gemma2` decoder block, please let me know if this works, just updated the branch on this as well, thanks again!

Comment (gobbleturk) [reply MaxText/maxtext_utils.py]: doesn't local/sliding attention use less flops than global attention? I think thats the main reason gemma2 uses sliding attention?

Comment (ZhaoyueCheng) [reply MaxText/maxtext_utils.py]: yeah haven't factored that in since we are using `dot_product` which still computes the full attention, will update to  `calculate_gemma2_flops` to include the sliding window attention flops calculation, thanks for the comment!

Comment (gobbleturk) [reply MaxText/maxtext_utils.py]: calculate_flops is meant to calculate the necessary flops for the model to train, not the amount of flops it actually uses. This allows it to be used to measure MFU

Comment (ZhaoyueCheng) [reply MaxText/maxtext_utils.py]: Thanks again for the details! updated the calculation to include sliding window attention for gemma2 and factor in the 2 stacked decoder layer in ffn layers

**Review by gobbleturk — APPROVED** [review]
Awesome, thank you for fixing the tflops calculation!

**Comment by ZhaoyueCheng** [comment]
> Awesome, thank you for fixing the tflops calculation!

Thanks for the detailed review and suggestions!!

**Comment by salrowili** [comment]
@ZhaoyueCheng I would like to thank you for adding Gemma 2 support which really is appreciated. 

I have pre-trained Gemma 2 using maxtext but now i stuck since i could not convert MaxText to hugging face format to do the SFT stage. Gemma2, unlike Llama, Gemma and Mistral, uses local and global attention . Is there any thing you can do to "MaxText/llama_or_mistral_ckpt.py" script to include Gemma 2 checkpoint conversion to HF format? The global and local attention used in the following weights: 
```
mlp_global
mlp_local
post_ffw_norm_global
post_ffw_norm_local
post_self_attention_norm_global
post_self_attention_norm_local
pre_ffw_norm_global
pre_ffw_norm_local
pre_self_attention_norm_global
pre_self_attention_norm_local
self_attention_global
```

This related to https://github.com/google/maxtext/issues/829
